loadbalancer/algorithm.go

```diff
@@ -197,9 +201,66 @@ func (ch *consistentHash) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 	return ctx.Route.LBEndpoints[choice]
 }
 
+type powerOfRandomNChoices struct {
+	rand            *rand.Rand
+	numberOfChoices int
+}
+
+// getScore returns negative value of inflightrequests count.
+func (p *powerOfRandomNChoices) getScore(e routing.LBEndpoint) int {
+	// endpoints with higher inflight request should have lower score
+	return -e.Metrics.GetInflightRequests()
+}
+
+// getRandomIndex returns random integer with the range of 0 and l
+// and skips excluded integers. If the length is equal to excluded length
+// it behaves as if excluded list is empty
+func (p *powerOfRandomNChoices) getRandomIndex(l int, exl map[int]struct{}) int {
+	if len(exl) == 0 || len(exl) == l {
+		return p.rand.Intn(l)
+	}
+
+	for {
+		r := p.rand.Intn(l)
+		if _, ok := exl[r]; !ok {
+			return r
+		}
+	}
+}
+
+// Apply implements routing.LBAlgorithm with power of random N choices algorithm.
+func (p *powerOfRandomNChoices) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	ne := len(ctx.Route.LBEndpoints)
+	exl := make(map[int]struct{}, p.numberOfChoices)
+
+	c := p.getRandomIndex(ne, exl)
+	best := ctx.Route.LBEndpoints[c]
+	exl[c] = struct{}{}
+
+	for i := 1; i < p.numberOfChoices; i++ {
+		newC := p.getRandomIndex(ne, exl)
+		ce := ctx.Route.LBEndpoints[newC]
+		exl[newC] = struct{}{}
+
+		if p.getScore(ce) > p.getScore(best) {
+			best = ce
+		}
+	}
+	return best
+}
+
+// newPowerOfRandomNChoices selects N random backends and picks the one with less outstanding requests.
+func newPowerOfRandomNChoices(endpoints []string) routing.LBAlgorithm {
+	t := time.Now().UnixNano()
+	return &powerOfRandomNChoices{
+		rand:            rand.New(rand.NewSource(t)),
```

Comment:
From https://golang.org/pkg/math/rand/
>  The default Source is safe for concurrent use by multiple goroutines, but Sources created by NewSource are not.

Reply:
This is the same implementation of newRandom.

Reply:
`newRandom` also has this problem unfortunately. 
> The default Source is safe for concurrent use by multiple goroutines, but Sources created by NewSource are not.

Top-level functions, such as Float64 and Int use mutex, see https://golang.org/src/math/rand/rand.go?#L382

Reply:
I think, it is better different algorithms managing their own custom random objects(instead using global rand source) and I agree about concurrent access should be safe. That's why, I created safeRand object to manage these locking operations and inject it into random and powerofrandomnchoices algorithms.

Reply:
IMO there is no benefit for each algorithm to have its own random over using global random unless algorithm uses one mutes to guard random and something else (like `roundRobin` does). 

See also my comments on why `roundRobin` and `consistentHash` do not need `safeRand`.

Reply:
After some thought I have to correct myself. It might make sense to use random+mutex instance per algorithm instance. This way only requests following the route would compete for the mutex as opposed to global random where requests from all routes (that use the same algorithm) compete for it. I guess that is the reason we see random instance per algorithm.

Reply:
yes random instance per algorithm instance makes sense

Reply:
I added rand and mutex into powerOfRandomNChoices algorithm like you proposal.